Lightcaster-Studios/Beamerman
# Implement Score System

Changes:
- (Unrelated to score) player death ghost now correctly reflects the player model
- (Unrelated to score) DefaultMap directional light no longer casts shadows (high performance hit, with little benefit)
- (Unrelated to score) Add delay in lazer destroy to prevent null exceptions
- (Unrelated to score) Refactor powerups to use parent's OnComponentOverlap
- Create middleman: ScoreMiddlemanFunctions to integrate scoretracker with various objects
- Create score system, with individual tracking of various score types (e.g. score from powerups, score from reflects, etc)
- HUD element to display total score
- Victory screen display total score/score breakdown

Todo:
- Die not due to other player's actions, steal score and distribute to all players [Separate PR]
- Detect damage done by barrel and use that to calculate score too [Separate PR]
- Detect object destroy and increment score (mainly for skyhook pillar) [Separate PR]
- HUD element to show score increments [Separate PR]

Current Bugs:
- Some Horizontal/Diagonal Mirrors do not trigger the score increase [Separate PR]
- Hitting players with protective temporary powerups still grants score (e.g. TempArmor) [Separate PR]

How to Test:
- Load ScoreTestMap
- Open IndividualScoreTracker blueprint
- Verify that all the variables under "ScoreType" are present under the "RecomputeScore" macro
- Test that each variable updates correctly using the objects in ScoreTestMap
- ScoreTrackers are exposed on the individual players, including the breakdown of each score. Use the World Outliner to access them
- Note: Currently ObjectDestroyScore has not been implemented properly

Score Breakdown (as of 9ad9271):
| Score Type                | Gain               |
|---------------------------|--------------------|
| Passive Gain              | 10 (per 2 seconds) |
| Player Hit                | 200                |
| Player Killed             | 500                |
| Player Kill Score Theft % | 15%                |
| Powerup Pickup            | 50                 |
| Point Orb Pickup          | 450                |
| ~~Lazer Mirror Reflect~~     | ~~20~~                 |
| ~~Lazer Prism Reflect~~       | ~~50~~                 |
| Crate Hit                 | 20                 |
| Explosive Barrel Hit      | 50                 |
| Skyhook Pillar Hit        | 30                 |

Reflection Bonus Score Scaling:
![image](https://user-images.githubusercontent.com/43085796/135113320-c4342fbc-d55f-47bb-995b-74160eafd838.png)
